sprockets/mixins/amqp/__init__.py

```diff
@@ -28,223 +31,481 @@ class PublishingMixin(object):
+
+    def _on_channel_closed(self, channel, reply_code, reply_text):
+        """Invoked by pika when RabbitMQ unexpectedly closes the channel.
+
+        Channels are usually closed if you attempt to do something that
+        violates the protocol, such as re-declare an exchange or queue with
+        different parameters.
+
+        In this case, we just want to log the error and create a new channel
+        after setting the state back to connecting.
+
+        :param pika.channel.Channel channel: The closed channel
+        :param int reply_code: The numeric reason the channel was closed
+        :param str reply_text: The text reason the channel was closed
+
+        """
+        LOGGER.warning('Channel %i was closed: (%s) %s',
+                       channel, reply_code, reply_text)
+        self._state = self.STATE_CONNECTING
+        self._channel = self._connection.channel(self._on_channel_open)
+        if self._on_unavailable:
+            self._on_unavailable(self)
 
     def _on_channel_open(self, channel):
-        """Called when the RabbitMQ accepts the channel open request.
+        """This method is invoked by pika when the channel has been opened.
+        The channel object is passed in so we can make use of it.
 
-        :param pika.channel.Channel channel: The channel opened with RabbitMQ
+        :param pika.channel.Channel channel: The channel object
 
         """
-        LOGGER.debug('RabbitMQ channel opened')
+        LOGGER.debug('Channel opened')
         self._channel = channel
+        self._state = self.STATE_CONNECTED
+        self._channel.add_on_flow_callback(self._on_channel_flow)
+        self._channel.add_on_return_callback(self._on_message_returned)
         self._channel.add_on_close_callback(self._on_channel_closed)
-        self._connecting = False
+        if self._on_ready:
+            self._on_ready(self)
         self._condition.notify_all()
 
-    def _on_channel_closed(self, channel, reply_code, reply_text):
-        """Called when the RabbitMQ accepts the channel close request.
+    def _on_channel_flow(self, method):
+        """When RabbitMQ indicates the connection is unblocked, set the state
+        appropriately and attempt to publish any deferred messages.
+
+        :param pika.spec.Channel.Flow method: The Channel flow frame
 
-        :param pika.channel.Channel channel: The channel closed with RabbitMQ
-        :param int reply_code: The code for the disconnect
-        :param str reply_text: The disconnect reason
+        """
+        if method.active:
+            LOGGER.warning('Channel flow enabled')
+            self._state = self.STATE_CONNECTED
+            if self._on_ready:
+                self._on_ready(self)
+        else:
+            LOGGER.info('Channel flow disabled')
+            self._state = self.STATE_BLOCKED
+            if self._on_unavailable:
+                self._on_unavailable(self)
+
+    @staticmethod
+    def _on_message_returned(*args, **_):
```

Comment:
If we're going to have this, then we should have the ability to register a callback that is invoked with the returned method.